### PR TITLE
Added RUN without saving routine

### PR DESCRIPTION
**Pull request by Kakise**
It was tested on my Ti-Nspire CX CAS (latest OS) with no trouble and works great with fireemu too. The only weird problem is that with both of your "official" MicroPython app, the calculator will freeze right after execution of the problem. I solved it using one of the micropython.tns files you sent on ti-planet (I guess, I lost the link): [micropython (2).zip](https://github.com/Vogtinator/pyWrite/files/3067741/micropython.2.zip)

### CODE REVIEW

File: editortask.cpp [hunk 1]
```diff
@@ -441,17 +441,28 @@ void EditorTask::menuSaveAs()
 }
 
 void EditorTask::menuRun()
-{
+{
```
Comment (Vogtinator) [inline editortask.cpp]: Whitespace error

File: editortask.cpp [hunk 2]
```diff
@@ -441,17 +441,28 @@ void EditorTask::menuSaveAs()
...
-    nl_exec(settings_task.mpython_path.content().c_str(), sizeof(argv), const_cast<char**>(argv));
+    {	
+		FILE *f = fopen("/documents/tmp.py", "wb");
+		if(f)
+		{
+			fwrite(buffer.c_str(), buffer.length(), 1, f);
+			fclose(f);
+		}
+		
+		const char* argv[] = {"/documents/tmp.py"};
```
Comment (Vogtinator) [inline editortask.cpp]: This part should be deduplicated by splitting it out in a separate method.

**Review by Vogtinator — CHANGES_REQUESTED** [review]
Approach looks good, just needs some code cleanup.

**Comment by Kakise** [comment]
Should be better now

File: editortask.cpp [hunk 3]
```diff
@@ -542,3 +535,11 @@ void EditorTask::changeSelection(unsigned int cursor_pos_new)
         sel_start = cursor_pos;
     }
 }
+
+void EditorTask::runFile(const char *file)
```
Comment (Kakise) [inline editortask.cpp:541]: Routine to run a file using Micropython

File: editortask.cpp [hunk 4]
```diff
@@ -441,27 +441,20 @@ void EditorTask::menuSaveAs()
...
-    {	
+    {
 		FILE *f = fopen("/documents/tmp.py", "wb");
 		if(f)
 		{
 			fwrite(buffer.c_str(), buffer.length(), 1, f);
 			fclose(f);
 		}
-		
-		const char* argv[] = {"/documents/tmp.py"};
```
Comment (Kakise) [inline editortask.cpp]: Deduplicated this part of the code

File: editortask.cpp [hunk 5]
```diff
@@ -444,14 +444,18 @@ void EditorTask::menuRun()
...
-        dialog_task.showMessage("You have to save the file first!");
-        return;
-    }
-
-    const char* argv[] = {filepath.c_str()};
-    key_hold_down = true;
-
-    nl_exec(settings_task.mpython_path.content().c_str(), sizeof(argv), const_cast<char**>(argv));
+		FILE *f = fopen("/documents/tmp.py", "wb");
+		if(f)
```
Comment (Vogtinator) [inline editortask.cpp]: Does  it still make sense to run the file if opening or writing fails?

File: editortask.h [hunk 6]
```diff
@@ -47,6 +47,7 @@ class EditorTask : public Task
     unsigned int linesUntil(const char *end);
     void updateCursor();
     void changeSelection(unsigned int cursor_pos_new);
+	void runFile(const char *file);
```
Comment (Vogtinator) [inline editortask.h]: Whitespace error

File: editortask.cpp (hunk 3 above)
Comment (Vogtinator) [reply editortask.cpp:541]: Looks good

**Review by Vogtinator — CHANGES_REQUESTED** [review]
Almost there!

File: editortask.h (hunk 6 above)
Comment (Kakise) [reply editortask.h]: Weird how I don't see those errors in my text editor :/

File: editortask.cpp (hunk 5 above)
Comment (Kakise) [reply editortask.cpp]: Nope, I'll add a message to inform the user if the save has failed, it should be better and would be consistent with the menuSaveAs() routine imho

Comment (Vogtinator) [reply editortask.cpp]: There was some `saveAndRun` code before, that might be useful here

Comment (Kakise) [reply editortask.cpp]: Yup, found it there: https://github.com/Vogtinator/pyWrite/commit/a204ee54ca4067d5c6ce9c5babe619ff8e65f49f#diff-677b969513424dad8457dcbb75529942L174 my code should be better thanks to your advices now

File: editortask.cpp [hunk 7]
```diff
@@ -444,14 +444,21 @@ void EditorTask::menuRun()
...
+			fwrite(buffer.c_str(), buffer.length(), 1, f);
+			fclose(f);
+			
+			runFile("/documents/tmp.py");
+			remove("/documents/tmp.py");
+		} else {
+			dialog_task.showMessage("Could not save 'tmp.py'!");
+		}
+
+    } else {
```
Comment (Vogtinator) [inline editortask.cpp]: Inconsistent style of bracing - here you can remove the {} completely, it's just a single statement

File: editortask.cpp [hunk 8]
```diff
@@ -444,14 +444,21 @@ void EditorTask::menuRun()
...
-    nl_exec(settings_task.mpython_path.content().c_str(), sizeof(argv), const_cast<char**>(argv));
+		FILE *f = fopen("/documents/tmp.py", "wb");
+		if(f)
+		{
+			fwrite(buffer.c_str(), buffer.length(), 1, f);
+			fclose(f);
+			
+			runFile("/documents/tmp.py");
+			remove("/documents/tmp.py");
+		} else {
```
Comment (Vogtinator) [inline editortask.cpp]: Inconsistent style of bracing - here you can remove the {} completely, it's just a single statement

**Review by Vogtinator — CHANGES_REQUESTED** [review]
Just style issues remaining

File: editortask.cpp [hunk 9]
```diff
@@ -444,14 +444,21 @@ void EditorTask::menuRun()
...
-    }
-
-    const char* argv[] = {filepath.c_str()};
-    key_hold_down = true;
-
-    nl_exec(settings_task.mpython_path.content().c_str(), sizeof(argv), const_cast<char**>(argv));
+		FILE *f = fopen("/documents/tmp.py", "wb");
+		if(f)
+		{
+			fwrite(buffer.c_str(), buffer.length(), 1, f);
```
Comment (Vogtinator) [inline editortask.cpp]: Actually, this should also check for write errors

**Review by Vogtinator — CHANGES_REQUESTED** [review]

File: editortask.cpp [hunk 10]
```diff
@@ -449,16 +449,19 @@ void EditorTask::menuRun()
 		{
 			fwrite(buffer.c_str(), buffer.length(), 1, f);
 			fclose(f);
-			
+		}
+		
+		if (!ferror(f))
```
Comment (Vogtinator) [inline editortask.cpp]: I'm not sure whether this is implemented fully in ndless - can you check the return value of fwrite instead?

File: editortask.cpp (hunk 9 above)
Comment (Kakise) [reply editortask.cpp]: Did this as it'll check for any error during closing and writing:
```suggestion
		if (!ferror(f))
		{
			runFile("/documents/tmp.py");
			remove("/documents/tmp.py");
		}
```

File: editortask.cpp (hunk 10 above)
Comment (Kakise) [reply editortask.cpp]: Hmm it worked fine when compiled with the latest ndless sdk on both my calculator and firemu but if you really want to, I can

Comment (Kakise) [reply editortask.cpp]:
```suggestion
				FILE *f = fopen("/documents/tmp.py", "wb");
		if(f)
		{
			if(fwrite(buffer.c_str(), buffer.length(), 1, f))
				runFile("/documents/tmp.py");
			else
			{
				dialog_task.showMessage("Could not save 'tmp.py'!");
				return;
			}
			fclose(f);
			remove("/documents/tmp.py");
		}
		else
			dialog_task.showMessage("Could not save 'tmp.py'!");
```

I don't know how to do it in a cleaner way tho...

Comment (Vogtinator) [reply editortask.cpp]: Now it doesn't close the file if `fwrite` fails, what about:

```
FILE *f = fopen("/documents/tmp.py", "wb");
if(f && fwrite(buffer.c_str(), buffer.length(), 1, f))
    runFile("/documents/tmp.py");
else
    dialog_task.showMessage("Could not save 'tmp.py'!");

if(f)
{
    fclose(f);
    remove("/documents/tmp.py");
}
```

Comment (Kakise) [reply editortask.cpp]: Way cleaner, thank you very much

**Review by Vogtinator — CHANGES_REQUESTED** [review]
Code looks good, but it's using tabs instead of four spaces for indentation - once that's fixed I'll merge it

**Comment by Kakise** [comment]
> Code looks good, but it's using tabs instead of four spaces for indentation - once that's fixed I'll merge it

Annnd, fixed !

**Comment by Vogtinator** [comment]
Annnd, merged!

Thanks for the contribution!

**Pull request merged**